### ISSUE TITLE
autodoc_default_options issues a warning on a missing special method

### ISSUE THREAD

**Issue by ynikitenko** [sphinx-doc/sphinx#6771]
**Describe the bug**
I want to document `__call__` methods of my classes when they exist.
I use such configuration in conf.py:
autodoc_default_options = {
    'members': True,
    'special-members': '`__call__`',
}
But when I launch that, it gives a warning:

WARNING: missing attribute mentioned in :members: or `__all__`: module lena.core, attribute `__call__`

**To Reproduce**
```
make html
```

**Expected behavior**
I think that defaults should not raise errors when they are not applicable.

**Environment info**
- OS: Fedora Core 29.
- Python version: 2.7.16
- Sphinx version: 1.8.5
- Sphinx extensions:  'sphinx.ext.autodoc',
    'sphinx.ext.autosummary',
    'sphinx.ext.coverage',
    'sphinx.ext.mathjax',
    'sphinx.ext.todo',
    'sphinx.ext.viewcode',
    'sphinxcontrib.napoleon'.

**Comment by tk0miya** [comment]
On my local, documents are build successfully with following files:
```
# example.py

class Foo:
    def __call__(self):
        pass
```
```
# index.rst
.. automodule:: example
```

Could you share your document or reproducible example please?

**Comment by ynikitenko** [comment]
Sorry, I don't think I can easily reproduce that. Did you try to make a separate module with `__init__.py` ? BTW, the problem seems to be not when there is a `__call__`, but when it's missing.

My module init it the following:

_lena/core/`__init__.py`_

`
"""Lena combines calculations using *sequences*.

*Sequences* consist of *elements*.
Basic Lena sequences and element types are defined in *lena.core* module.
"""
from lena.core.exceptions import (
    LenaException, 
    LenaAttributeError, LenaEnvironmentError, LenaIndexError, LenaKeyError,
    LenaRuntimeError, LenaStopFill, LenaTypeError, LenaValueError
)
from lena.core.lena_sequence import LenaSequence

from lena.core.adapters import Call, FillCompute, FillInto, FillRequest, Run
from lena.core.sequence import Sequence
from lena.core.source import Source
from lena.core.fill_compute_seq import FillComputeSeq
from lena.core.fill import Fill
from lena.core.fill_request_seq import FillRequestSeq
from lena.core.split import Split

`__all__` = [
    'Call', 'FillCompute', 'FillInto', 'FillRequest', 'Run',
    'LenaSequence', 'Sequence', 'Source', 'Split',
    'Fill',
    'FillComputeSeq', 'FillRequestSeq',
    'LenaException',
    'LenaAttributeError',
    'LenaEnvironmentError', 
    'LenaIndexError', 'LenaKeyError',
    'LenaRuntimeError', 'LenaStopFill', 'LenaTypeError', 
    'LenaValueError',
]
`

**Comment by tk0miya** [comment]
Please share your project or make a reproducible example. It is hard to fix the problem without reproducing error. And it is hard to reproduce it from partial information. I don't know about your program; `lena.core` at all.

**Comment by ynikitenko** [comment]
Dear Takeshi, 
I tried to reproduce the example. I attach that project.
In the beginning sphinx repeated the warning (but then it disappeared, I don't know why). Try it yourself.

```
docs]$make html
Running Sphinx v1.8.5
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: [] 0 added, 1 changed, 0 removed
reading sources... [100%] index                                                        
WARNING: missing attribute mentioned in :members: or __all__: module main.example, attribute __call__
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                         
generating indices... genindex py-modindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in build/html.
```

[example.tar.gz](https://github.com/sphinx-doc/sphinx/files/3799217/example.tar.gz)

**Comment by tk0miya** [comment]
Thank you for example. I confirmed the warning. And it seems this project is configured as following:
```
autodoc_default_options = {
     # 'members': True,
     'special-members': '__call__',
     # 'undoc-members': True,
     # 'exclude-members': '__weakref__'
}
```

Is this intended? This means automodule should show a member named `__call__` only. I think the warning message is correct.

**Comment by ynikitenko** [comment]
When I uncomment members option, I have an error:

> $ make html
Running Sphinx v1.8.5
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index                                                        
Exception occurred:
  File "/usr/lib64/python2.7/site-packages/sphinx/ext/autodoc/__init__.py", line 82, in members_option
    return [x.strip() for x in arg.split(',')]
AttributeError: 'bool' object has no attribute 'split'
The full traceback has been saved in /tmp/sphinx-err-xMWsEJ.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:19: html] Error 2

In the real project I use a separate option

> autodoc_default_flags = ['members']

When I use this option with uncommented 'members' in this project, there is no error.

To sum up, this is what I think:

1. Two different options for same thing should be deprecated. I think a warning would be fine for that.
2. If there is an option in `autodoc_default_options`, it should work (or it should be documented which combination of options won't work).
3. I think that the default option should give no warning. This is a whole project option. Imagine one has very many classes, and wants to make `call` method documented by default. How one can accomplish that without a warning, if a few of the classes don't have this method (but many do)? Also, if there are no members, `members` default option probably gives no warning, so it is would be more consistent.

**Comment by tk0miya** [comment]
Ah, sorry for confusing you. `autodoc_default_options` is a successor of `autodoc_default_flags`. `autodoc_default_options` can take an argument to the option. Now `autodoc_default_flags` has been marked as deprecated. It will be warned since Sphinx-2.x (and removed in Sphinx-4.0.0).

>How one can accomplish that without a warning, if a few of the classes don't have this method (but many do)? 

How about this?
```
autodoc_default_options = {
    'members': True,
    'special-members': '__call__',
}
```
With `member` option, everything goes well to me. Is there any reason not to add it?

**Comment by tk0miya** [comment]
Now I'm closing this now because this is not a bug of Sphinx.
Please feel free to reopen if you still have a problem.
Thanks

**Comment by ynikitenko** [comment]
Thanks for reminding.

Now all my code is published, and I can still reproduce the error.
However, I can't reopen this issue. Could you please do it yourself?

The working example of failure:

    git clone https://github.com/ynikitenko/lena
    cd lena/docs
    make html
    # it should work fine

Now (update: comment autodoc_default_flags = ['members']) and uncomment autodoc options, as proposed earlier, and set them to 

    autodoc_default_options = {
        'members': True,
        'special-members': '__call__',
    }

    vi source/conf.py
    make html
    # and here are warnings:
    WARNING: missing attribute mentioned in :members: or __all__: module lena.core, attribute __call__
    WARNING: missing attribute mentioned in :members: or __all__: module lena.flow.group_plots, attribute __call__

These warnings were missing when I didn't use ``autodoc_default_options``. Of course, a module (e.g. ``lena.core``) should not have a ``__call__`` method. I don't know why it warns about ``lena.flow.group_plots``, but, as I said, there was no warning before. And the warning itself (missing default attribute?) is wrong: the default attribute should be default when available, but if it is missing, it should not be an error.

In docs/source/core.rst there is a line

    .. automodule:: lena.core.adapters
    :special-members: __call__

but it should not raise an error (or the error must be specific to this module, not to whole ``lena.core``). Otherwise I've no idea of the source of these errors.

I have to write for each of my classes with this call method

    :special-members: __call__

because I can't use default options. What would you suggest? I'd rather not ignore warnings. Maybe there is a special option to ignore specific warnings?

**Comment by ynikitenko** [comment]
Maybe I should add that after commenting again ``autodoc_default_options``, several warnings appear:

    WARNING: missing attribute mentioned in :members: or __all__: module lena.core.adapters, attribute __call__
    WARNING: missing attribute mentioned in :members: or __all__: module lena.flow.iterators, attribute __call__
    WARNING: missing attribute mentioned in :members: or __all__: module lena.variables.variable, attribute __call__
    WARNING: missing attribute mentioned in :members: or __all__: module lena.variables.variable, attribute __getitem__

They disappear, though, when I run ``make html`` several times.
@tk0miya hope you receive these notifications.

**Comment by tk0miya** [comment]
Reopened now.

**Comment by tk0miya** [comment]
It seems you're using `:no-members:` option.
```
.. automodule:: lena.core
    :no-members:
```
https://github.com/ynikitenko/lena/blob/master/docs/source/core.rst

With combination of `autodoc_default_options`, this is interpreted as following because `:member:` option is canceled by the `:no-members:` option for the automodule directive.
```
.. automodule:: lena.core
   :special-members: __call__
```
Is this intended?

The warnings are emitted on the specified special member is not found on target object except `:members:` option without argument is given.
```
.. automodule:: lena.core
   :members:
   :special-members: __call__

   This is not warned because :members: option is given.
   It scans all of members of the lena.core module because of :member: option.

.. automodule:: lena.core
   :special-members: __call__

   This is warned because of the lena.core module does not have __call__ attribute.
   The automodule directive determines target attributes using :special-members: option.

.. automodule:: lena.core
   :members: Sequence
   :special-members: __call__

   This is also warned. The automodule directive determines target attributes using both
   :members: and :special-members: option.
```

Please let me know what are you going with `:no-members:` option. I think it causes your problem. In my understanding, it is meaningless.

**Comment by ynikitenko** [comment]
Unfortunately, I can't see the examples on the link you sent.
There is a class

    .. autoclass:: Source
        :members:
        :special-members: __call__

And I use both ``:members:`` and ``:special-members:`` with that.
Also there is a module

    .. automodule:: lena.core.adapters
        :special-members: __call__

but it doesn't prohibit the members of this module. 

I use ``:no-members:`` for the following: I want to include the docstring of the module from ``__init__.py`` (and use ``:automodule:`` for that), but I want to manually include only the classes I specify explicitly (maybe it's a bad practice and I should change that).

Many thanks that you looked into my code and suggested this hint. Unfortunately I don't understand well what you say exactly (I can't find the citations you mention).

**Comment by tk0miya** [comment]
Only I mentioned is an automodule directive with `:no-members:` option.
```
.. automodule:: lena.core
    :no-members:
```

In your case, `:members:` and `:special-members: __call__` options are given from `autodoc_default_options`. But `:member:` option is canceled because the directive has `:no-members:` option also. As a result, it is recognized internally as following:
```
.. automodule:: lena.core
   :special-members: __call__
```

This is as same as I commented at https://github.com/sphinx-doc/sphinx/issues/6771#issuecomment-551091356

How about adding `:no-special-members:` to this call? I think It would work.
```
.. automodule:: lena.core
    :no-members:
    :no-special-members:
```

**Comment by ynikitenko** [comment]
Now I understand, thank you. I didn't think about this distinction. I also thought that the ``:members:`` directive for a module relates only to module contents, not its class members.

In [autodoc docs](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directives) it's written:

> If you want to automatically document members, there’s a members option:
> .. automodule:: noodle
>       :members:
> will document all module members (recursively)

IMHO, it should be reformulated somehow if that relates also to class members (because they are described separately below).

Yes, this fix really works great! Many thanks for your time and advice!

Probably you may close this issue or fix the docs or do what you think is appropriate.

**Comment by tk0miya** [comment]
Good to hear. About docs, I have a plan to rewrite it if I have time. So I'm closing this.
